test/parallel/test-fs-rm.js

```diff
@@ -287,74 +286,125 @@ function removeAsync(dir) {
   // IBMi has a different access permission mechanism
   // This test should not be run as `root`
   if (!common.isIBMi && (common.isWindows || process.getuid() !== 0)) {
-    function makeDirectoryReadOnly(dir, mode) {
-      let accessErrorCode = 'EACCES';
-      if (common.isWindows) {
-        accessErrorCode = 'EPERM';
-        execSync(`icacls ${dir} /deny "everyone:(OI)(CI)(DE,DC)"`);
-      } else {
-        fs.chmodSync(dir, mode);
+    // Check that deleting a file that cannot be accessed using rmsync throws:
+    // https://github.com/nodejs/node/issues/38683
+    {
+      function isValidState(exists, err) {
+        // On Windows, we are allowed to access and modify the contents of a
+        // read-only folder.
+        return common.isWindows ?
+          (exists === false && err === null) :
+          (exists === true && err?.code === 'EACCES');
       }
-      return accessErrorCode;
-    }
 
-    function makeDirectoryWritable(dir) {
-      if (fs.existsSync(dir)) {
-        if (common.isWindows) {
-          execSync(`icacls ${dir} /remove:d "everyone"`);
-        } else {
-          fs.chmodSync(dir, 0o777);
+      {
+        const dirname = nextDirPath();
+        const filePath = path.join(dirname, 'text.txt');
+
+        fs.mkdirSync(dirname, { recursive: true });
+        fs.writeFileSync(filePath, 'hello');
+
+        fs.chmodSync(filePath, 0o444);
+        fs.chmodSync(dirname, 0o444);
+
+        let err = null;
+
+        try {
+          fs.rmSync(filePath, { force: true });
+        } catch (_err) {
+          err = _err;
+        }
+
+        try { fs.chmodSync(dirname, 0o777); } catch {}
+        try { fs.chmodSync(filePath, 0o777); } catch {}
+
+        if (!isValidState(fs.existsSync(filePath), err)) {
+          throw err;
```

Comment:
`err` could potentially be `null` here:
```suggestion
          throw err || new Error('Invalid state');
```
We could use the [Error Cause proposal](https://github.com/tc39/proposal-error-cause) for a clearer stack trace:
```suggestion
          const e = new Error('Invalid state');
          e.cause = err;
          throw e;
```